### PR TITLE
Add support for Zcash unified addresses - memory optimized

### PR DESCRIPTION
**Pull request by jarys**
specification: https://zips.z.cash/zip-0316
optimization: f4jumble now uses C blake2b module

### CODE REVIEW

**Comment by jarys** [comment]
This PR is ready for review. I don't understand why CI `core fw btconly build` fails. There is no error message.

File: core/src/apps/zcash/f4jumble.py
Comment (prusnak) [inline core/src/apps/zcash/f4jumble.py:38]: Why 22?

Upstream says the min length is 48 (and max length is 4194368): https://github.com/zcash/librustzcash/blob/37fc28634e811fdf8db9274a0080cff17c6b6a09/components/f4jumble/src/lib.rs#L67-L70

File: core/src/apps/zcash/unified_addresses.py
Comment (prusnak) [inline core/src/apps/zcash/unified_addresses.py:30]:
```suggestion
    if typecode in  [Typecode.P2PKH, Typecode.P2SH]:
        return 20
    if typecode in [Typecode.SAPLING, Typecode.ORCHARD]:
        return 43
```

File: core/src/apps/zcash/unified_addresses.py
Comment (prusnak) [inline core/src/apps/zcash/unified_addresses.py:45]: uPy supports only UTF-8

```suggestion
    return bytes(hrp) + bytes(16 - len(hrp))
```

**Review by prusnak — CHANGES_REQUESTED** [review]

**Comment by andrewkozlik** [comment]
> I don't understand why CI `core fw btconly build` fails. There is no error message.

At the end of the log you can see
```
o%trezor/enums/ZcashReceiverTypecode.py
trezor/enums/ZcashReceiverTypecode.py
```
This is the output of `tools/check-bitcoin-only`. We should clarify the error message.

If you run the following commands you will get the same result.
```
BITCOIN_ONLY=1 make build_firmware
strings core/build/firmware/firmware.bin | grep Zcash
```

I think what you need to do is look at `SConscript.firmware` and `SConscript.unix` and exclude `trezor/enums/Zcash*.py` from the Bitcoin-only build.

File: core/src/apps/zcash/f4jumble.py
Comment (jarys) [reply core/src/apps/zcash/f4jumble.py:38]: You are right.

I started by copy-pasting implementation by daira. Then I found that it's outdated and I forgot to change this.

https://github.com/daira/jumble/blob/main/f4jumble.py#L18

Comment (jarys) [reply core/src/apps/zcash/f4jumble.py]: https://github.com/trezor/trezor-firmware/pull/2398/commits/8dd8eff3e1e4c1fa0593987a0d3b697bcd5aed0a

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py]: https://github.com/trezor/trezor-firmware/pull/2398/commits/e54626dcad3d7fa22fc85c337bb44a760a024d99

**Comment by jarys** [comment]
Thank you for clarification Andrew. I'm removing `trezor/enums/Zcash*.py` from btconly build in https://github.com/trezor/trezor-firmware/pull/2398/commits/2a18731b99d1ea9d8c30404f1efc95f629bb6c10 .

**Review by prusnak — APPROVED** [review]

**Comment by jarys** [comment]
may I squash and merge , or waiting for another review ? @andrewkozlik

File: core/src/apps/zcash/unified_addresses.py
Comment (matejcik) [inline core/src/apps/zcash/unified_addresses.py:44]: please use py3.10-style annotations, i.e., in this case `dict[Typecode, bytes]` without importing `Dict`

File: core/src/apps/zcash/unified_addresses.py
Comment (matejcik) [inline core/src/apps/zcash/unified_addresses.py:23]: nit: these should be tuples `(Typecode.P2PKH, Typecode.P2SH)` not lists

**Review by matejcik — APPROVED** [review]
did a brief review, two small nitpicks.

@andrewkozlik do we want to add this to the upcoming release?

**Comment by andrewkozlik** [comment]
> may I squash and merge , or waiting for another review ?

I will try to review today.

> do we want to add this to the upcoming release?

Is there any reason not to?

**Comment by matejcik** [comment]
> Is there any reason not to?

the freeze is today; my question is, are you as the main reviewer confident enough in this code to include it?

File: common/protob/messages-zcash.proto
Comment (andrewkozlik) [inline common/protob/messages-zcash.proto:17]: Why do we need this `enum` in a protobuf definition? As far as I am able to tell this is never communicated over USB, so it could be defined more locally in `unified_addresses.py`. Will it be needed in the future in the protobuf messages? If yes, then how exactly?

File: core/tests/test_apps.zcash.unified_addresses.py
Comment (andrewkozlik) [inline core/tests/test_apps.zcash.unified_addresses.py:7]: I found some official-looking test vectors at https://github.com/zcash-hackworks/zcash-test-vectors/blob/master/test-vectors/json/unified_address.json. It looks like the ones in this file are different. Do these also come from some official source?

File: core/src/apps/zcash/unified_addresses.py
Comment (andrewkozlik) [inline core/src/apps/zcash/unified_addresses.py:117]: > Consumers MUST reject Unified Addresses/Viewing Keys [...] that include both P2SH and P2PKH Transparent Addresses, or that contain only a Transparent Address.

Technically we should be checking this and ideally have a unit test for each of these cases.

File: core/src/apps/zcash/unified_addresses.py
Comment (andrewkozlik) [inline core/src/apps/zcash/unified_addresses.py:46]: Also
```python
assert Typecode.SAPLING in receivers or Typecode.ORCHARD in receivers
```
But I am wondering whether it is right to treat this with `assert`. It would seem more appropriate to use `if ... raise` at least until we have some guarantee that the calling code is checking this. (Right now `encode()` is not used.)

File: common/protob/messages-zcash.proto
Comment (andrewkozlik) [reply common/protob/messages-zcash.proto:17]: The only place that I can think of where this could be used is for unified addresses in `GetAddress` if there would be a repeated field for the receivers. In that case I wonder if it wouldn't be better to recycle `InputScriptType` for this purpose in `GetAddress`.

**Comment by andrewkozlik** [comment]
> the freeze is today; my question is, are you as the main reviewer confident enough in this code to include it?

I don't think we need to force this into the release, since Suite integration of unified addresses is not on the roadmap yet. So I am not sure this could be used in production by anyone in the nearest future.

On the other hand if we happen to merge this today somehow, then it's fine to include this in the freeze. (My apologies for not getting this reviewed earlier. I had a lot on my plate.)

**Comment by prusnak** [comment]
Let's merge after the freeze if there is no strong reason to merge it before.

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py]: [fcc5fd1](https://github.com/trezor/trezor-firmware/pull/2398/commits/fcc5fd1cccbb92be5248f9179479e32e5856bb77)

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py]: [c536e71](https://github.com/trezor/trezor-firmware/pull/2398/commits/c536e71da57fdbcde00ff0c87405f3f51b4a25bf)

File: common/protob/messages-zcash.proto
Comment (jarys) [reply common/protob/messages-zcash.proto:17]: No. This is not exposed and it will never be exposed. `t_address_n` and `z_address_n` fields will be used for getting unified addresses.

I will move this enum into unified_addresses.py

File: core/tests/test_apps.zcash.unified_addresses.py
Comment (jarys) [reply core/tests/test_apps.zcash.unified_addresses.py:7]: I generated test vector by the same official library (zcash-test-vectors) with two minor changes:

- I generated ALL possible combinations of receivers. (The generated 10 random combinations.)
- I serialized test vectors differently to load them more easily in python.

Is it ok?

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py:115]: adding the check [c7b9b7d](https://github.com/trezor/trezor-firmware/pull/2398/commits/c7b9b7d3954062246288f47b6dac644da0ee82b2)

Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py:115]: adding the second check [e780855](https://github.com/trezor/trezor-firmware/pull/2398/commits/e78085529fcc47d53fcf7045947f24d02837dcfc)

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py:44]: `assert` seems to fine to me since input of this function will be always created internally.

Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py:44]: [6a410a8](https://github.com/trezor/trezor-firmware/pull/2398/commits/6a410a833c9ba28c91a0b364a7856b7fd08a5cbf)

File: core/tests/test_apps.zcash.unified_addresses.py
Comment (andrewkozlik) [reply core/tests/test_apps.zcash.unified_addresses.py:7]: > - I generated ALL possible combinations of receivers. (They generated 10 random combinations.)

Generating more combinations is certainly good work. It would be nicer though if our test vectors were a superset of the official ones.

> - I serialized test vectors differently to load them more easily in python.

Absolutely OK.

File: common/protob/messages-zcash.proto
Comment (jarys) [reply common/protob/messages-zcash.proto]: [5f33136](https://github.com/trezor/trezor-firmware/pull/2398/commits/5f33136b275176497f1e72d4651beb1aa3ce7d1d)

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py:130]: fixing typechecking [163f493](https://github.com/trezor/trezor-firmware/pull/2398/commits/163f493d46ecf31c7f10077a1a9b102145ad6cae)

File: core/tests/test_apps.zcash.unified_addresses.py
Comment (jarys) [reply core/tests/test_apps.zcash.unified_addresses.py:7]: updated to official test vectors [163f493](https://github.com/trezor/trezor-firmware/pull/2398/commits/163f493d46ecf31c7f10077a1a9b102145ad6cae)

Comment (jarys) [reply core/tests/test_apps.zcash.unified_addresses.py:7]: While testing on new testvectors I found a mistake :-( I reversed receivers order (which should be ascending). It is repaired now. I got confused by receivers priority order (which is descending).

Comment (jarys) [reply core/tests/test_apps.zcash.unified_addresses.py]: [42f4215](https://github.com/trezor/trezor-firmware/pull/2398/commits/42f42155148673156a5275cead794d5ee18cba12)

File: core/src/apps/zcash/unified_addresses.py
Comment (andrewkozlik) [inline core/src/apps/zcash/unified_addresses.py:64]: nit: I'd simplify this for better legibility, since it's only to estimate the size for buffer preallocation.
```suggestion
    length = 16  # 16 bytes for padding
    for receiver_bytes in receivers.values():
        length += 2  # typecode (1 byte) + length (1 byte)
        length += len(receiver_bytes)
```

**Review by andrewkozlik — APPROVED** [review]
utACK

**Comment by matejcik** [comment]
looks like this has all the approval it needs
please squash the fixups, rebase on current master, and fix the UI fixtures conflict

File: core/src/apps/zcash/unified_addresses.py
Comment (jarys) [reply core/src/apps/zcash/unified_addresses.py]: [b70a2c1](https://github.com/trezor/trezor-firmware/pull/2398/commits/b70a2c15d742251515ceb49d4b3a96a88cdf2d16)

**Review by andrewkozlik — APPROVED** [review]

**Comment by jarys** [comment]
core fw regular debug build:
```
region `FLASH2' overflowed by 1024 bytes
```
:cry:

**Comment by andrewkozlik** [comment]
Also, the address in `test_unified_address` has `Invalid receivers order`.

> region `FLASH2' overflowed by 1024 bytes

Strange, I still have 4 kB left when I build with `PYOPT=0 poetry run make -C core build_firmware` on my machine.

```
code_length: 1696256
-rw-rw-r-- 1 andrew andrew  1699840 Aug  9 10:14 firmware.bin
-rwxrwxr-x 1 andrew andrew   786432 Aug  9 10:14 firmware.bin.p1
-rwxrwxr-x 1 andrew andrew   913408 Aug  9 10:14 firmware.bin.p2
```

**Pull request merged**
